ntoskrnl/io/pnpmgr/pnpmgr.c

```diff
@@ -1781,11 +1861,16 @@ IopQueryHardwareIds(PDEVICE_NODE DeviceNode,
                               &Stack);
    if (NT_SUCCESS(Status))
    {
-      /*
-       * FIXME: Check for valid characters, if there is invalid characters
-       * then bugcheck.
-       */
+      IsValideID = IopValidateID((PWCHAR)IoStatusBlock.Information,
+                                 MAX_DEVICE_ID_LEN,
+                                 MAX_SEPARATORS_MULTISZ,
+                                 TRUE);
+
+      if (!IsValideID)
+         DPRINT1("Invalid HardwareIDs. DeviceNode - %p\n", DeviceNode);
```

Comment:
If this case is supposed to bugcheck, then by all means do so. Or at least assert.

Reply:
I do not see that Windows used BugCheck when checking IDs. Only IoWriteErrorLogEntry(). I do not think we should stop the system because of the wrong ID. A driver with the wrong ID can only be third-party. We will check "our" drivers and do not allow the use of incorrect IDs. For example, in usbhub_new, we need to add a terminating NULL-terminator for MultiSz (and usbport, too). I sent you a patch "17-Fixing_QueryIDs_in_usport _ & _ usbhub.patch".

Reply:
You did a test on Windows and generate invalid IDs and Windows did not bugcheck?
Because, again according to https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/irp-mn-query-id (and to the WinDDK doc you screenshotted above), "If a driver returns an ID with an illegal character, the system will bug check."